lib/teleterm/gateway/gateway.go

```diff
@@ -32,19 +32,46 @@ import (
 	"github.com/sirupsen/logrus"
 )
 
-// New creates an instance of Gateway
+// New creates an instance of Gateway. It starts a listener on the specified port but it doesn't
+// start the proxy – that's the job of Serve.
 func New(cfg Config) (*Gateway, error) {
 	if err := cfg.CheckAndSetDefaults(); err != nil {
 		return nil, trace.Wrap(err)
 	}
 
-	listener, err := net.Listen("tcp", fmt.Sprintf("%s:%s", cfg.LocalAddress, cfg.LocalPort))
+	result, err := newListenerAndLocalProxy(cfg, cfg.LocalPort)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	cfg.LocalPort = result.LocalPort
+
+	gateway := &Gateway{
+		cfg:          &cfg,
+		closeContext: result.closeContext,
+		closeCancel:  result.closeCancel,
+		localProxy:   result.localProxy,
+	}
+
+	return gateway, nil
+}
+
+type newListenerAndLocalProxyResult struct {
+	LocalPort    string
+	listener     net.Listener
+	closeContext context.Context
+	closeCancel  context.CancelFunc
+	localProxy   *alpn.LocalProxy
+}
+
+func newListenerAndLocalProxy(cfg Config, port string) (*newListenerAndLocalProxyResult, error) {
```

Comment:
I'm actually not sure what's the common approach in Go when I need to return more than two values from a method. I opted for creating this struct specifically for this purpose but I'm open to suggestions.

Reply:
I think normally I would make this `Gateway.init` function to assign directly but I understand it will not work for the swap case.

Not sure if letting `Gateway` own this new struct will be better than holding each piece separately. something like:
```
type Gateway struct {
	cfg   *Config
        proxy *gatewayProxy
}
type gatewayProxy struct {
        *alpn.LocalProxy
	localPort    string
	closeContext context.Context
	closeCancel  context.CancelFunc
}
```

Reply:
The only downside to that solution would be that we'd have `Config.LocalPort` and `gatewayProxy.localPort`. I think I'd rather keep those fields as they are right now so that we can avoid having do deal with that duplication.

Reply:
IMHO, it's better to have them separate. `Config.LocalPort` is the static input and shouldn't be overwritten. `gatewayProxy.localPort` is the real deal and is what should be returned by `gateway.LocalPort`. But that's just me. I am Ok with current approach. Let's if others have other suggestions.